src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyRenderer.cs

```diff
@@ -22,12 +22,21 @@ namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
 internal sealed partial class WebAssemblyRenderer : WebRenderer
 {
     private readonly ILogger _logger;
+    private readonly Dispatcher _dispatcher;
 
     public WebAssemblyRenderer(IServiceProvider serviceProvider, ILoggerFactory loggerFactory, JSComponentInterop jsComponentInterop)
         : base(serviceProvider, loggerFactory, DefaultWebAssemblyJSRuntime.Instance.ReadJsonSerializerOptions(), jsComponentInterop)
     {
         _logger = loggerFactory.CreateLogger<WebAssemblyRenderer>();
 
+        // if SynchronizationContext.Current is null, it means we are on the single-threaded runtime
+        // otherwise capture the JSSynchronizationContext from the main thread, which runtime already installed.
+        // if user installed SynchronizationContext different from JSSynchronizationContext, they need to make sure the behavior is consistent with JSSynchronizationContext.
+        var currentCtx = SynchronizationContext.Current;
```

Comment:
Instead of capturing the context here, it might make more sense to capture the context earlier during the program execution and flow it here. Some options:
* ModuleInitializer: Might be too son.
* WebAssemblyHostbuilder: This is pretty much the first line in the program for a blazor app, so there's a very narrow window for the user to override this.

I don't think we care about supporting a custom synchronization context that is not the JS one.

Reply:
Added `WebAssemblyHostBuilder.InitializeWebAssemblyRenderer`